Sources/StreamChat/Controllers/ChannelListController/ChannelListController_Tests.swift

```diff
@@ -420,13 +473,19 @@ private class TestEnvironment {
 // A concrete `ChannelListControllerDelegate` implementation allowing capturing the delegate calls
 private class TestDelegate: QueueAwareDelegate, ChatChannelListControllerDelegate {
     @Atomic var state: DataController.State?
+    @Atomic var willChangeChannels_called = false
     @Atomic var didChangeChannels_changes: [ListChange<ChatChannel>]?
-    
+
     func controller(_ controller: DataController, didChangeState state: DataController.State) {
         self.state = state
         validateQueue()
     }
-    
+
+    func controllerWillChangeChannels(_ controller: ChatChannelListController) {
+        willChangeChannels_called = true
+        validateQueue()
+    }
```

Comment:
Just for a consideration, I usually prefer to create a closure that allows to define what will be executed in function so for this case I'd have something like this:

```swift
var controllerWillChangeChannelsBody: (ChatChannelListController) -> Void = { _ in }

func controllerWillChangeChannels(_ controller: ChatChannelListController) {
    controllerWillChangeChannelsBody(controller)
    validateQueue() // not sure if it would come here or to the default body
}
```

This approach is a bit more verbose but generally gives you flexibility in tests. But as I mentioned, just for future considerations.

Generally it is annoying to switch to this approach, especially with 1100 tests, but it is quite simple and fast to use it in new tests.

Reply:
@olejnjak I see the benefits of this approach in terms of better flexibility, on the other hand, it also requires bigger setup in the test:

```swift
controller.simulateSomething()
AssertAsync.willBeTrue(mock.willChangeChannels_called)

vs.

var wasCalled = false
mock.controllerWillChangeChannelsBody = { wasCalled = true  }

controller.simulateSomething()
AssertAsync.willBeTrue(wasCalled)
```

It's quite common for our current test setup to have multiple "mocks" for the same object but with different functionalities. So instead of having one almighty global mock, we have multiple local mocks. This basically replicates what you suggest but with concrete types rather than closures.

In this comparison, I don't see any clear winner there. However, thank you for pointing this pattern out. I'm sure there are cases where it works better than the "hardcoded" mock style.